quinn-proto/src/token.rs

```diff
@@ -19,20 +19,26 @@ use crate::{
 // - AEAD nonce is always set to 0
 // in other words, for each ticket, use different key derived from random using HKDF
 
-pub struct RetryToken {
+pub struct RetryToken<'a> {
     /// The destination connection ID set in the very first packet from the client
     pub orig_dst_cid: ConnectionId,
     /// The time at which this token was issued
     pub issued: SystemTime,
+    /// Random bytes for deriving AEAD key
+    pub random_bytes: &'a [u8],
 }
 
-impl RetryToken {
+impl<'a> RetryToken<'a> {
+    pub const RANDOM_BYTES_LEN: usize = 32;
+
     pub fn encode(
         &self,
-        key: &impl HmacKey,
+        key: &impl PseudoRandomKey,
         address: &SocketAddr,
         retry_src_cid: &ConnectionId,
     ) -> Vec<u8> {
+        let aead_key: ring::aead::LessSafeKey = AeadKey::from_hkdf(key, self.random_bytes);
```

Comment:
I would like to know how to do this better i.e. I know that `endpoint.rs` knows the actual type i.e. `S::AeadKey` but I am not sure how to use that here?

Reply:
Would it make sense for `encode` and `from_bytes` to take an explicit `K: AeadKey` type parameter?

Reply:
I think you'll have to use an associated type in either `PseudoRandomKey` or `AeadKey` to link one to the other. In this case, I think it might make sense to add a method to `PseudoRandomKey` that constructs an `AeadKey` (rather than the other way around), and then `PseudoRandomKey` can also have an associated type for the type of the `AeadKey` it builds.

We don't usually use the `impl Trait` in argument position, instead using named bound parameters, I think it would make sense to do so here, too (so give `RetryToken` a `P` argument?).

Reply:
I'm actually a fan of argument position `impl Trait` for the concision when a type parameter never needs to be explicitly specified/otherwise qualified, though I don't mind much either way.

Reply:
@djc  As far as I can understand your comment it seems that you are suggesting the following,

```rust
pub trait PseudoRandomKey: Send + Sized + Sync {
    type AeadKey: AeadKey;
    fn aead_from_hkdf(&self) -> Self::AeadKey;
    ...
}
```

```rust
impl crypto::PseudoRandomKey for hkdf::Prk {
    type AeadKey = ring::aead::LessSafeKey;

    fn aead_from_hkdf(&self) -> Self::AeadKey {
        ...
    }
}
```

then I can just do `let aead_key = key.aead_from_hkdf(self.random_bytes);`

@Ralith  and you suggested the following,

```rust
pub fn encode<K: AeadKey, P: PseudoRandomKey>(
    &self,
    key: &P,
    address: &SocketAddr,
    retry_src_cid: &ConnectionId,
) -> Vec<u8>
```

which can then be invoked from `endpoint.rs` 

```rust
let token = RetryToken {
    orig_dst_cid: dst_cid,
    issued: SystemTime::now(),
    random_bytes: &random_bytes,
}
.encode::<S::AeadKey, S::PseudoRandomKey>(
    &*server_config.token_key,
    &remote,
    &temp_loc_cid,
);
```

based on the discussion it seems that we would prefer the first approach right?

Reply:
With the first approach you'd still need a type parameter, but only one instead of two, and it wouldn't need to be passed explicitly. That seems fine to me.

Reply:
Yup, sounds good to me.